src/Microsoft.AspNetCore.Routing/Matchers/DefaultMatchProcessorFactory.cs

```diff
@@ -20,145 +20,120 @@ internal class DefaultMatchProcessorFactory : MatchProcessorFactory
             _serviceProvider = serviceProvider;
         }
 
-        public override MatchProcessor Create(MatchProcessorReference matchProcessorReference)
+        public override MatchProcessor Create(string parameterName, IRouteConstraint value, bool optional)
         {
-            if (matchProcessorReference == null)
+            if (value == null)
             {
-                throw new ArgumentNullException(nameof(matchProcessorReference));
+                throw new ArgumentNullException(nameof(value));
             }
 
-            if (matchProcessorReference.MatchProcessor != null)
+            return InitializeMatchProcessor(parameterName, optional, value, null);
+        }
+
+        public override MatchProcessor Create(string parameterName, MatchProcessor value, bool optional)
+        {
+            if (value == null)
+            {
+                throw new ArgumentNullException(nameof(value));
+            }
+
+            return InitializeMatchProcessor(parameterName, optional, value, null);
+        }
+
+        public override MatchProcessor Create(string parameterName, string value, bool optional)
+        {
+            if (value == null)
             {
-                return matchProcessorReference.MatchProcessor;
+                throw new ArgumentNullException(nameof(value));
             }
 
             // Example:
             // {productId:regex(\d+)}
             //
             // ParameterName: productId
-            // ConstraintText: regex(\d+)
-            // ConstraintName: regex
-            // ConstraintArgument: \d+
-
-            (var constraintName, var constraintArgument) = Parse(matchProcessorReference.ConstraintText);
+            // value: regex(\d+)
+            // name: regex
+            // argument: \d+
+            (var name, var argument) = Parse(value);
 
-            if (!_options.ConstraintMap.TryGetValue(constraintName, out var constraintType))
+            if (!_options.ConstraintMap.TryGetValue(name, out var type))
```

Comment:
type -> constraintType

Reply:
It's pretty clear right? What other type would it get confused with?

Reply:
sure. probably should have `nit:`ed it 😄